File: acp-sdk-go/identity.go
package acp

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"strings"
	"time"

	"github.com/google/uuid"
)

type AgentIdentity struct {
	AgentID              string `json:"agent_id"`
	SigningPrivateKey    string `json:"signing_private_key"`
	SigningPublicKey     string `json:"signing_public_key"`
	EncryptionPrivateKey string `json:"encryption_private_key"`
	EncryptionPublicKey  string `json:"encryption_public_key"`
	SigningKID           string `json:"signing_kid"`
	EncryptionKID        string `json:"encryption_kid"`
}

type AgentIDParts struct {
	Name   string
	Domain string
}

type IdentityBundle struct {
	Identity         AgentIdentity
	IdentityDocument map[string]any
}

type ProviderIdentityInput struct {
	AgentID              string
	SigningPrivateKey    string
	EncryptionPrivateKey string
	SigningPublicKey     string
	EncryptionPublicKey  string
	SigningKID           string
	EncryptionKID        string
}

const (
	identityFileName    = "identity.json"
	identityDocFileName = "identity_document.json"
)

var agentIDPattern = regexp.MustCompile(`^agent:(?P<name>[^@]+)(?:@(?P<domain>.+))?$`)

func ParseAgentID(agentID string) (AgentIDParts, error) {
	matches := agentIDPattern.FindStringSubmatch(agentID)
	if len(matches) == 0 {
		return AgentIDParts{}, ValidationError(fmt.Sprintf("Invalid agent identifier: %s", agentID))
	}
	parts := AgentIDParts{}
	for index, name := range agentIDPattern.SubexpNames() {
		switch name {
		case "name":
			parts.Name = matches[index]
		case "domain":
			parts.Domain = matches[index]
		}
	}
	if strings.TrimSpace(parts.Name) == "" {
		return AgentIDParts{}, ValidationError(fmt.Sprintf("Invalid agent identifier: %s", agentID))
	}
	return parts, nil
}

func SanitizeAgentID(agentID string) string {
	var out strings.Builder
	for _, char := range agentID {
		if (char >= 'a' && char <= 'z') ||
			(char >= 'A' && char <= 'Z') ||
			(char >= '0' && char <= '9') ||
			char == '.' || char == '_' || char == '-' {
			out.WriteRune(char)
			continue
		}
		out.WriteRune('_')
	}
	return out.String()
}

func identityPath(storageDir string, agentID string) string {
	return filepath.Join(storageDir, SanitizeAgentID(agentID))
}

func CreateIdentity(agentID string) (AgentIdentity, error) {
	if _, err := ParseAgentID(agentID); err != nil {
		return AgentIdentity{}, err
	}
	signingPrivateKey, signingPublicKey, err := GenerateEd25519Keypair()
	if err != nil {
		return AgentIdentity{}, err
	}
	encryptionPrivateKey, encryptionPublicKey, err := GenerateX25519Keypair()
	if err != nil {
		return AgentIdentity{}, err
	}
	return AgentIdentity{
		AgentID:              agentID,
		SigningPrivateKey:    signingPrivateKey,
		SigningPublicKey:     signingPublicKey,
		EncryptionPrivateKey: encryptionPrivateKey,
		EncryptionPublicKey:  encryptionPublicKey,
		SigningKID:           "sig-" + strings.ReplaceAll(uuid.NewString(), "-", "")[:12],
		EncryptionKID:        "enc-" + strings.ReplaceAll(uuid.NewString(), "-", "")[:12],
	}, nil
}

type BuildIdentityDocumentInput struct {
	Identity             AgentIdentity
	DirectEndpoint       string
	RelayHints           []string
	TrustProfile         string
	Capabilities         map[string]any
	ValidDays            int
	AMQPService          map[string]any
	MQTTService          map[string]any
	HTTPSecurityProfile  string
	RelaySecurityProfile string
}

func BuildIdentityDocument(input BuildIdentityDocumentInput) (map[string]any, error) {
	if !IsSupportedTrustProfile(input.TrustProfile) {
		return nil, ValidationError(fmt.Sprintf("Unsupported trust profile: %s", input.TrustProfile))
	}
	now := time.Now().UTC()
	validUntil := now.Add(time.Duration(maxInt(input.ValidDays, 1)) * 24 * time.Hour)
	service := map[string]any{
		"direct_endpoint": nullableString(input.DirectEndpoint),
		"relay_hints":     append([]string{}, input.RelayHints...),
	}
	if input.AMQPService != nil {
		service["amqp"] = input.AMQPService
	}
	if input.MQTTService != nil {
		service["mqtt"] = input.MQTTService
	}
	if strings.TrimSpace(input.DirectEndpoint) != "" && strings.TrimSpace(input.HTTPSecurityProfile) != "" {
		service["http"] = map[string]any{
			"endpoint":         strings.TrimSpace(input.DirectEndpoint),
			"security_profile": strings.TrimSpace(input.HTTPSecurityProfile),
		}
	}
	if len(input.RelayHints) > 0 && strings.TrimSpace(input.RelaySecurityProfile) != "" {
		service["relay"] = map[string]any{
			"endpoint":         strings.TrimSpace(input.RelayHints[0]),
			"security_profile": strings.TrimSpace(input.RelaySecurityProfile),
		}
	}
	document := map[string]any{
		"acp_identity_version": ACPIdentityVersion,
		"agent_id":             input.Identity.AgentID,
		"created_at":           now.Format(time.RFC3339),
		"valid_until":          validUntil.Format(time.RFC3339),
		"trust_profile":        input.TrustProfile,
		"keys": map[string]any{
			"signing": map[string]any{
				"kid":        input.Identity.SigningKID,
				"alg":        "Ed25519",
				"public_key": input.Identity.SigningPublicKey,
			},
			"encryption": map[string]any{
				"kid":        input.Identity.EncryptionKID,
				"alg":        "X25519",
				"public_key": input.Identity.EncryptionPublicKey,
			},
		},
		"service":      service,
		"capabilities": mapOrDefault(input.Capabilities),
	}
	signatureInput, err := CanonicalJSONBytes(document)
	if err != nil {
		return nil, err
	}
	signatureValue, err := SignBytes(signatureInput, input.Identity.SigningPrivateKey)
	if err != nil {
		return nil, err
	}
	document["signature"] = map[string]any{
		"algorithm": "Ed25519",
		"signed_by": input.Identity.SigningKID,
		"value":     signatureValue,
	}
	return document, nil
}

func VerifyIdentityDocument(identityDocument map[string]any) bool {
	requiredFields := []string{"agent_id", "keys", "service", "signature", "valid_until"}
	for _, field := range requiredFields {
		if _, ok := identityDocument[field]; !ok {
			return false
		}
	}
	trustProfile, ok := identityDocument["trust_profile"].(string)
	if !ok || !IsSupportedTrustProfile(trustProfile) {
		return false
	}
	validUntilRaw, ok := identityDocument["valid_until"].(string)
	if !ok {
		return false
	}
	validUntil, err := time.Parse(time.RFC3339, validUntilRaw)
	if err != nil || !validUntil.After(time.Now().UTC()) {
		return false
	}
	keys, ok := identityDocument["keys"].(map[string]any)
	if !ok {
		return false
	}
	signing, ok := keys["signing"].(map[string]any)
	if !ok {
		return false
	}
	signingPublicKey, ok := signing["public_key"].(string)
	if !ok || strings.TrimSpace(signingPublicKey) == "" {
		return false
	}
	signature, ok := identityDocument["signature"].(map[string]any)
	if !ok {
		return false
	}
	signatureValue, ok := signature["value"].(string)
	if !ok || strings.TrimSpace(signatureValue) == "" {
		return false
	}
	unsigned := copyMap(identityDocument)
	delete(unsigned, "signature")
	signatureInput, err := CanonicalJSONBytes(unsigned)
	if err != nil {
		return false
	}
	return VerifySignature(signatureInput, signatureValue, signingPublicKey)
}

func WriteIdentity(storageDir string, identity AgentIdentity, identityDocument map[string]any) error {
	path := identityPath(storageDir, identity.AgentID)
	if err := os.MkdirAll(path, 0o755); err != nil {
		return ValidationError(fmt.Sprintf("unable to create identity directory: %v", err))
	}
	identityBytes, err := json.MarshalIndent(identity, "", "  ")
	if err != nil {
		return ValidationError(fmt.Sprintf("unable to serialize identity: %v", err))
	}
	if err := os.WriteFile(filepath.Join(path, identityFileName), identityBytes, 0o600); err != nil {
		return ValidationError(fmt.Sprintf("unable to persist identity: %v", err))
	}
	documentBytes, err := json.MarshalIndent(identityDocument, "", "  ")
	if err != nil {
		return ValidationError(fmt.Sprintf("unable to serialize identity document: %v", err))
	}
	if err := os.WriteFile(filepath.Join(path, identityDocFileName), documentBytes, 0o644); err != nil {
		return ValidationError(fmt.Sprintf("unable to persist identity document: %v", err))
	}
	return nil
}

func ReadIdentity(storageDir string, agentID string) (*IdentityBundle, error) {
	path := identityPath(storageDir, agentID)
	identityBytes, err := os.ReadFile(filepath.Join(path, identityFileName))
	if err != nil {
		return nil, nil
	}
	documentBytes, err := os.ReadFile(filepath.Join(path, identityDocFileName))
	if err != nil {
		return nil, nil
	}
	var identity AgentIdentity
	if err := json.Unmarshal(identityBytes, &identity); err != nil {
		return nil, ValidationError(fmt.Sprintf("invalid stored identity: %v", err))
	}
	document, err := ParseJSONMap(documentBytes)
	if err != nil {
		return nil, err
	}
	return &IdentityBundle{
		Identity:         identity,
		IdentityDocument: document,
	}, nil
}

func IdentityFromProvider(input ProviderIdentityInput) (AgentIdentity, error) {
	signingPublicKey := strings.TrimSpace(input.SigningPublicKey)
	if signingPublicKey == "" {
		derived, err := Ed25519PublicFromPrivate(input.SigningPrivateKey)
		if err != nil {
			return AgentIdentity{}, err
		}
		signingPublicKey = derived
	}
	encryptionPublicKey := strings.TrimSpace(input.EncryptionPublicKey)
	if encryptionPublicKey == "" {
		derived, err := X25519PublicFromPrivate(input.EncryptionPrivateKey)
		if err != nil {
			return AgentIdentity{}, err
		}
		encryptionPublicKey = derived
	}
	signingKID := strings.TrimSpace(input.SigningKID)
	if signingKID == "" {
		signingKID = "sig-" + strings.ReplaceAll(uuid.NewString(), "-", "")[:12]
	}
	encryptionKID := strings.TrimSpace(input.EncryptionKID)
	if encryptionKID == "" {
		encryptionKID = "enc-" + strings.ReplaceAll(uuid.NewString(), "-", "")[:12]
	}
	return AgentIdentity{
		AgentID:              input.AgentID,
		SigningPrivateKey:    input.SigningPrivateKey,
		SigningPublicKey:     signingPublicKey,
		EncryptionPrivateKey: input.EncryptionPrivateKey,
		EncryptionPublicKey:  encryptionPublicKey,
		SigningKID:           signingKID,
		EncryptionKID:        encryptionKID,
	}, nil
}

func copyMap(input map[string]any) map[string]any {
	output := map[string]any{}
	for key, value := range input {
		output[key] = value
	}
	return output
}

func mapOrDefault(input map[string]any) map[string]any {
	if input == nil {
		return map[string]any{}
	}
	return input
}

File: sdks/typescript/src/identity.ts
/*
 * Copyright 2026 ACP Project
 * Licensed under the Apache License, Version 2.0
 * See LICENSE file for details.
 */

import { mkdirSync, readFileSync, writeFileSync } from "node:fs";
import { join } from "node:path";
import { randomUUID } from "node:crypto";
import {
  ACP_IDENTITY_VERSION,
  isSupportedTrustProfile
} from "./constants";
import {
  ed25519PublicFromPrivate,
  generateEd25519Keypair,
  generateX25519Keypair,
  signBytes,
  verifySignature,
  x25519PublicFromPrivate
} from "./crypto";
import { JsonMap, JsonValue, canonicalJsonBytes, parseJsonMap } from "./jsonSupport";
import { validationError } from "./errors";

export interface AgentIdentity {
  agent_id: string;
  signing_private_key: string;
  signing_public_key: string;
  encryption_private_key: string;
  encryption_public_key: string;
  signing_kid: string;
  encryption_kid: string;
}

export interface AgentIdParts {
  name: string;
  domain?: string;
}

export interface IdentityBundle {
  identity: AgentIdentity;
  identity_document: JsonMap;
}

const IDENTITY_FILE_NAME = "identity.json";
const IDENTITY_DOC_FILE_NAME = "identity_document.json";

export function parseAgentId(agentId: string): AgentIdParts {
  const match = /^agent:(?<name>[^@]+)(?:@(?<domain>.+))?$/.exec(agentId);
  if (!match?.groups?.name) {
    throw validationError(`Invalid agent identifier: ${agentId}`);
  }
  return {
    name: match.groups.name,
    domain: match.groups.domain || undefined
  };
}

export function sanitizeAgentId(agentId: string): string {
  return agentId
    .split("")
    .map((ch) => (/^[A-Za-z0-9._-]$/.test(ch) ? ch : "_"))
    .join("");
}

function identityPath(storageDir: string, agentId: string): string {
  return join(storageDir, sanitizeAgentId(agentId));
}

export function createIdentity(agentId: string): AgentIdentity {
  parseAgentId(agentId);
  const signing = generateEd25519Keypair();
  const encryption = generateX25519Keypair();
  return {
    agent_id: agentId,
    signing_private_key: signing.private_key,
    signing_public_key: signing.public_key,
    encryption_private_key: encryption.private_key,
    encryption_public_key: encryption.public_key,
    signing_kid: `sig-${randomUUID().replace(/-/g, "").slice(0, 12)}`,
    encryption_kid: `enc-${randomUUID().replace(/-/g, "").slice(0, 12)}`
  };
}

export function buildIdentityDocument(input: {
  identity: AgentIdentity;
  direct_endpoint?: string;
  relay_hints: string[];
  trust_profile: string;
  capabilities?: JsonMap;
  valid_days: number;
  amqp_service?: JsonMap;
  mqtt_service?: JsonMap;
  http_security_profile?: string;
  relay_security_profile?: string;
}): JsonMap {
  if (!isSupportedTrustProfile(input.trust_profile)) {
    throw validationError(`Unsupported trust profile: ${input.trust_profile}`);
  }
  const now = new Date();
  const validUntil = new Date(now.getTime() + Math.max(1, input.valid_days) * 24 * 60 * 60 * 1000);
  const service: JsonMap = {
    direct_endpoint: input.direct_endpoint ?? null,
    relay_hints: input.relay_hints
  };
  if (input.amqp_service) {
    service.amqp = input.amqp_service;
  }
  if (input.mqtt_service) {
    service.mqtt = input.mqtt_service;
  }
  if (input.direct_endpoint && input.http_security_profile) {
    service.http = {
      endpoint: input.direct_endpoint,
      security_profile: input.http_security_profile
    };
  }
  if (input.relay_hints[0] && input.relay_security_profile) {
    service.relay = {
      endpoint: input.relay_hints[0],
      security_profile: input.relay_security_profile
    };
  }
  const document: JsonMap = {
    acp_identity_version: ACP_IDENTITY_VERSION,
    agent_id: input.identity.agent_id,
    created_at: now.toISOString(),
    valid_until: validUntil.toISOString(),
    trust_profile: input.trust_profile,
    keys: {
      signing: {
        kid: input.identity.signing_kid,
        alg: "Ed25519",
        public_key: input.identity.signing_public_key
      },
      encryption: {
        kid: input.identity.encryption_kid,
        alg: "X25519",
        public_key: input.identity.encryption_public_key
      }
    },
    service,
    capabilities: input.capabilities ?? {}
  };
  const signature = signBytes(
    canonicalJsonBytes(document as JsonValue),
    input.identity.signing_private_key
  );
  document.signature = {
    algorithm: "Ed25519",
    signed_by: input.identity.signing_kid,
    value: signature
  };
  return document;
}

export function verifyIdentityDocument(identityDocument: JsonMap): boolean {
  for (const field of ["agent_id", "keys", "service", "signature", "valid_until"]) {
    if (!(field in identityDocument)) {
      return false;
    }
  }
  if (
    typeof identityDocument.trust_profile !== "string" ||
    !isSupportedTrustProfile(identityDocument.trust_profile)
  ) {
    return false;
  }
  if (typeof identityDocument.valid_until !== "string") {
    return false;
  }
  const expires = Date.parse(identityDocument.valid_until);
  if (Number.isNaN(expires) || expires <= Date.now()) {
    return false;
  }
  const signature = identityDocument.signature as JsonMap | undefined;
  const signingPublicKey = ((((identityDocument.keys as JsonMap).signing as JsonMap).public_key ??
    "") as string)
    .trim();
  const signatureValue = ((signature?.value ?? "") as string).trim();
  if (!signingPublicKey || !signatureValue) {
    return false;
  }
  const unsigned = { ...identityDocument };
  delete unsigned.signature;
  return verifySignature(canonicalJsonBytes(unsigned as JsonValue), signatureValue, signingPublicKey);
}

export function writeIdentity(storageDir: string, identity: AgentIdentity, identityDocument: JsonMap): void {
  const path = identityPath(storageDir, identity.agent_id);
  mkdirSync(path, { recursive: true });
  writeFileSync(join(path, IDENTITY_FILE_NAME), JSON.stringify(identity, null, 2), "utf-8");
  writeFileSync(join(path, IDENTITY_DOC_FILE_NAME), JSON.stringify(identityDocument, null, 2), "utf-8");
}

export function readIdentity(storageDir: string, agentId: string): IdentityBundle | undefined {
  const path = identityPath(storageDir, agentId);
  try {
    const identity = JSON.parse(readFileSync(join(path, IDENTITY_FILE_NAME), "utf-8")) as AgentIdentity;
    const identityDocument = parseJsonMap(readFileSync(join(path, IDENTITY_DOC_FILE_NAME), "utf-8"));
    return { identity, identity_document: identityDocument };
  } catch {
    return undefined;
  }
}

export function identityFromProvider(input: {
  agent_id: string;
  signing_private_key: string;
  encryption_private_key: string;
  signing_public_key?: string;
  encryption_public_key?: string;
  signing_kid?: string;
  encryption_kid?: string;
}): AgentIdentity {
  const signingPublicKey =
    input.signing_public_key ?? ed25519PublicFromPrivate(input.signing_private_key);
  const encryptionPublicKey =
    input.encryption_public_key ?? x25519PublicFromPrivate(input.encryption_private_key);
  return {
    agent_id: input.agent_id,
    signing_private_key: input.signing_private_key,
    signing_public_key: signingPublicKey,
    encryption_private_key: input.encryption_private_key,
    encryption_public_key: encryptionPublicKey,
    signing_kid: input.signing_kid ?? `sig-${randomUUID().replace(/-/g, "").slice(0, 12)}`,
    encryption_kid: input.encryption_kid ?? `enc-${randomUUID().replace(/-/g, "").slice(0, 12)}`
  };
}
